Debug / Cheat: Debug Warp Menu (and Simple Warps for Cheat Menu)

Dumbledork01 opened the pull request:
While the console does support warping to entrances, it still requires looking up the hexadecimal values of each entrance. So, this PR is meant to fix that. 

**Debug Currently supports:**
* All warps defined in the OOT source code. (With the exception of warps that crash the game or warps that repeat other warps.)
* Grottos unique to their location (as in, you'll get the loot for that specific grotto when you warp there.) Exits also work correctly.
* Buttons to switch whether it will be day or night and whether you'll be an adult or child.
* Warps for shops, fairy fountains, and minigames. 
* Custom warp setting

**Cheat Currently supports:**
* Warping to most significant overworld locations.
* Warping to any dungeon.
* Buttons to switch whether it will be day or night and whether you'll be an adult or child.

Demonstration: https://youtu.be/O0QXOVv73Ao
Demonstration of debug warp menu: https://youtu.be/-H6RA-qiV68

**Cheat menu example:**
![Cheat Warp Menu](https://user-images.githubusercontent.com/77412657/185526856-ff607b27-bd8c-4c5d-9da3-3854454f6eee.png)

**Debug Menu Example:**
![Screenshot 2022-08-20 155202](https://user-images.githubusercontent.com/77412657/185767241-fa8a8c26-dadd-4563-8caa-de37d216d9bb.png)

Dumbledork01 committed:
Added basis for warp menu

Dumbledork01 committed:
Basis for warping logic added

garrettjoecox commented:
@JakeEdvenson a few things:

A) I would check out `soh/src/overlays/gamestates/ovl_select/z_select.c` for a good starting point, and maybe also using names from `soh/soh/util.cpp`, as we already have these names duplicated in more places than we should.

B) You should be able to achieve this with a single set of CVars rather than a CVar per warp.

C) This will especially be more useful than the built in warp menu for grottos, so if we can figure out how to either warp to each of them correctly, or warp to right outside of each grotto that would be great.

stratomaster64 commented:
Also keep in mind:
- Entrance values that would be input into the console command expect something from the entrance table, which is quite large due to accounting for age, time of day, and cutscene setups. You may wish to consider adding in some options for the first two since the entrance adjusts accordingly.
- For grottos (`z_door_ana`), you may want to look at the [entrance table](https://github.com/HarbourMasters/Shipwright/blob/develop-zhora/soh/src/overlays/actors/ovl_Door_Ana/z_door_ana.c#L55-L58) for where they could lead to. 
- Note also that grottos work by setting the `respawn` correlating to `RESPAWN_MODE_RETURN` to where you entered previously. Failing to do so will spawn Link at deku tree at (0,0,0) when he exits, and will softlock.
That being said, godspeed.

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into Simple-Warp-Cheat

Dumbledork01 committed:
Added Warp Logic for Grottos

Changed Warp Logic

now, sceneNames reflect the names found in soh/src/overlays/gamestates/ovl_select/z_select.c

Updated Warp Logic

Warps now include all main locations including dungeons

Logical Checks for Warps added

Updated logic again

Updated warp gui method

Removed old logic

Added grottos into warp logic

Added Generic Fairy Fountain Warp

Fixed Formatting

Fixed typo

Dumbledork01 commented:
@garrettjoecox Could you check if the logic I just pushed is more in line with what you're talking about? This was my understanding of what you're looking for. If I've got this right, then I'll start working on grottos. 
@stratomaster64 I think that the warps automatically update if you're adult or child. Anytime I warp as adult link, it automatically loads me into adult link's area (same goes for time of day). And if I warp as child link, it loads me into child link's area. I could make it so you can choose which to go to and it switches your age if you want me to, but I have a PR open that creates a cheat that does will switch your age at the click of a button.

garrettjoecox commented:
> @garrettjoecox Could you check if the logic I just pushed is more in line with what you're talking about? 

If you’re meaning from my point B) not quite, you should just be able to do something like CVarSet32(“gCheatWarpTo”, warpId) or similar, and then only need to check that one Cvar and determine where to warp to, rather than having a cvar per location and looping through them all to check where you should warp to

garrettjoecox commented:
Also one better, IIRC most of the functions in C are exposed to C++ with #include \<functions\> or similar, so no need for a cVar at all, you should just be able to call a warp function directly.

Sorry I’m not super descriptive, typing these out from my phone at work right now!

Dumbledork01 commented:
@garrettjoecox Thats okay! I appreciate the help! I've had some issues linking the C files to the C++ files (I have experience with both languages, but linking them is new to me,) so I'll probably start by testing this out with CVar and then go from there. Thanks for your help!

Dumbledork01 commented:
I left out unique Fairy Grottos for now since I figure they aren't super necessary for testing. I can add them in if we think they need to be there, otherwise there is a generic warp to a fairy fountain grotto. Also, currently you respawn whereever you were prior to warping. However, the logic works so that each grotto is it's own unique variant so they can be used for testing randomizers or whatever else needs testing.

Dumbledork01 committed:
Added shops for debugging shopsanity

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into Simple-Warp-Cheat

Dumbledork01 committed:
Fixed Linker Errors

Fixed bug in code

Fix Linker Error

Shouldn't have included "z_play.c," I'm assuming it's already included in globals or something.

Redid code to remove all linker errors

Fixed logic to prevent linker errors

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into Simple-Warp-Cheat

Dumbledork01 committed:
Refactored Code, Added features

Can search for keywords. Supports finding both locations and categories (Kokiri Forest: From Lost Woods and Kokiri Forest (all warps) can be accessed by typing in Kokiri Forest)
Created Debug Warps
Moving all the debug warps and grottos to the SaveEditor
Added debug warps
Added Warps to Debug Menu
Warps have been added to debug menu to help with going to specific locations including grottos, shops, and most areas that have specific entrance indexes.
Fixed Typo
Added more warps, added index table
Re-factored code to support searching
Made a structure that contains main areas, entrances within those areas, and their respective hexadecimal values. See Minigames, Fairy Fountains, and Shops for how this can be used to create a search function.
Fixed typo, added custom warps

Dumbledork01 committed:
Refactored Code, Added features

Can search for keywords. Supports finding both locations and categories (Kokiri Forest: From Lost Woods and Kokiri Forest (all warps) can be accessed by typing in Kokiri Forest)
Created Debug Warps
Moving all the debug warps and grottos to the SaveEditor
Added debug warps
Added Warps to Debug Menu
Warps have been added to debug menu to help with going to specific locations including grottos, shops, and most areas that have specific entrance indexes.
Fixed Typo
Added more warps, added index table
Re-factored code to support searching
Made a structure that contains main areas, entrances within those areas, and their respective hexadecimal values. See Minigames, Fairy Fountains, and Shops for how this can be used to create a search function.
Fixed typo, added custom warps
Custom Warps Save After Closing SoH

Dumbledork01 committed:
Custom Warps Save After Closing SoH

Dumbledork01 committed:
Merge branch 'Simple-Warp-Cheat' of https://github.com/JakeEdvenson/Shipwright into Simple-Warp-Cheat

Dumbledork01 committed:
Updated all logic for grottos

Added logic for grottos

Fixed typo

Dumbledork01 committed:
Merge branch 'HarbourMasters:develop-zhora' into Simple-Warp-Cheat

garrettjoecox commented:
First off, this was hugely instrumental in #1479 , so credit where credit is due you definitely did a bunch of helpful work here. Thanks a bunch 😄 

Secondly, I'm going to close this for now, as I'm not sure how much interest there is left on having the warps in ImGui when this is mostly taken care of now with the better debug warp screen. If you or anyone else feels otherwise feel free to @ me here or in discord and we can discuss further.

Dumbledork01 commented:
Glad it was helpful! @garrettjoecox
